src/dataflow/src/render/reduce.rs

```diff
@@ -126,25 +124,29 @@ where
                         // If any error occurs we produce both the error as output,
                         // but also a `Datum::Null` value to avoid causing the later
                         // "ReduceCollation" operator to panic due to absent aggregates.
-                        if results.is_empty() {
-                            let key = row_packer.finish_and_reuse();
-                            for aggr in aggregates_clone.iter() {
-                                match aggr.expr.eval(&datums_local, &temp_storage) {
-                                    Ok(val) => {
-                                        row_packer.push(val);
-                                    }
-                                    Err(e) => {
-                                        row_packer.push(Datum::Null);
-                                        results.push(Err(e.into()));
-                                    }
+                        let key = row_packer.finish_and_reuse();
+                        for aggr in aggregates_clone.iter() {
+                            match aggr.expr.eval(&datums_local, &temp_storage) {
+                                Ok(val) => {
+                                    row_packer.push(val);
+                                }
+                                Err(e) => {
+                                    return Some(Err(e.into()));
                                 }
                             }
-                            let row = row_packer.finish_and_reuse();
-                            results.push(Ok((key, row)));
                         }
                         datums = repurpose_allocation(datums_local);
-                        // Return accumulated results.
-                        results
+
+                        // Mint the final row, ideally re-using resources.
+                        use timely::communication::message::RefOrMut;
```

Comment:
Can we wrap this whole block into a function? it seems like we might be using this pattern again in a bunch of places in the rendering code

Reply:
We could. I was tempted to wait a bit before doing this to get a sense for how pervasive we wanted e.g. `RefOrMut` to be (it is a helper struct in `timely::communication` rather than some glorious type). Like, probably the "most pleasant" thing is a method on `RowPacker` that takes a `RefOrMut`, but this also leaks that type all over the place and might be gross.

Dunno. Inclined to punt for now until we see a few more uses. I do expect that packing in to a `&mut Row` will be a thing, but I'm not sure if we see `RefOrMut` other than in `flat_map_ref` calls.